File: benchmarks/python/deriv.py
#!/usr/bin/env python3

from __future__ import annotations

import sys

STEPS = 10


def val(n: int):
    return ("val", n)


def var(n: int):
    return ("var", n)


def add_expr(left, right):
    return ("add", left, right)


def mul_expr(left, right):
    return ("mul", left, right)


def pow_expr(left, right):
    return ("pow", left, right)


def ln_expr(expr):
    return ("ln", expr)


def pown(a: int, n: int) -> int:
    if n == 0:
        return 1
    if n == 1:
        return a
    b = pown(a, n // 2)
    return b * b * (1 if n % 2 == 0 else a)


def add(left, right):
    if left[0] == "val":
        n = left[1]
        if right[0] == "val":
            return val(n + right[1])
        if n == 0:
            return right
        if right[0] == "add":
            _, inner_left, inner_right = right
            if inner_left[0] == "val":
                return add(val(n + inner_left[1]), inner_right)
            if inner_right[0] == "val":
                return add(val(n + inner_right[1]), inner_left)
        return add_expr(left, right)
    if left[0] == "add":
        _, inner_left, inner_right = left
        return add(inner_left, add(inner_right, right))
    if right[0] == "val" and right[1] == 0:
        return left
    if right[0] == "val":
        return add(right, left)
    if right[0] == "add":
        _, inner_left, inner_right = right
        if inner_left[0] == "val":
            return add(val(inner_left[1]), add(left, inner_right))
        return add_expr(left, right)
    return add_expr(left, right)


def mul(left, right):
    if left[0] == "val":
        n = left[1]
        if right[0] == "val":
            return val(n * right[1])
        if n == 0:
            return val(0)
        if n == 1:
            return right
        if right[0] == "mul":
            _, inner_left, inner_right = right
            if inner_left[0] == "val":
                return mul(val(n * inner_left[1]), inner_right)
            if inner_right[0] == "val":
                return mul(val(n * inner_right[1]), inner_left)
        return mul_expr(left, right)
    if left[0] == "mul":
        _, inner_left, inner_right = left
        return mul(inner_left, mul(inner_right, right))
    if right[0] == "val" and right[1] == 0:
        return val(0)
    if right[0] == "val" and right[1] == 1:
        return left
    if right[0] == "val":
        return mul(right, left)
    if right[0] == "mul":
        _, inner_left, inner_right = right
        if inner_left[0] == "val":
            return mul(val(inner_left[1]), mul(left, inner_right))
        return mul_expr(left, right)
    return mul_expr(left, right)


def pow_(left, right):
    if left[0] == "val":
        m = left[1]
        if right[0] == "val":
            return val(pown(m, right[1]))
        if m == 0:
            return val(0)
        if right[0] == "val" and right[1] == 0:
            return val(1)
        if right[0] == "val" and right[1] == 1:
            return left
        return pow_expr(left, right)
    if right[0] == "val" and right[1] == 0:
        return val(1)
    if right[0] == "val" and right[1] == 1:
        return left
    if left[0] == "val" and left[1] == 0:
        return val(0)
    return pow_expr(left, right)


def ln(expr):
    if expr[0] == "val" and expr[1] == 1:
        return val(0)
    return ln_expr(expr)


def deriv(x: int, expr):
    tag = expr[0]
    if tag == "val":
        return val(0)
    if tag == "var":
        return val(1 if expr[1] == x else 0)
    if tag == "add":
        _, left, right = expr
        return add(deriv(x, left), deriv(x, right))
    if tag == "mul":
        _, left, right = expr
        return add(mul(left, deriv(x, right)), mul(right, deriv(x, left)))
    if tag == "pow":
        _, left, right = expr
        return mul(
            pow_(left, right),
            add(
                mul(mul(right, deriv(x, left)), pow_(left, val(-1))),
                mul(ln(left), deriv(x, right)),
            ),
        )
    _, inner = expr
    return mul(deriv(x, inner), pow_(inner, val(-1)))


def count(expr) -> int:
    tag = expr[0]
    if tag in ("val", "var"):
        return 1
    if tag == "ln":
        return count(expr[1])
    _, left, right = expr
    return count(left) + count(right)


def main() -> int:
    sys.setrecursionlimit(200000)
    expr = pow_(var(1), var(1))
    for step in range(STEPS):
        expr = deriv(1, expr)
        print(f"{step + 1} count: {count(expr)}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: benchmarks/rust/deriv.rs
#[derive(Clone)]
enum Expr {
    Val(i64),
    Var(i64),
    Add(Box<Expr>, Box<Expr>),
    Mul(Box<Expr>, Box<Expr>),
    Pow(Box<Expr>, Box<Expr>),
    Ln(Box<Expr>),
}

const STEPS: usize = 10;

fn pown(a: i64, n: i64) -> i64 {
    if n == 0 {
        1
    } else if n == 1 {
        a
    } else {
        let b = pown(a, n / 2);
        let factor = if n % 2 == 0 { 1 } else { a };
        b * b * factor
    }
}

fn add(f: Expr, g: Expr) -> Expr {
    match f {
        Expr::Val(n) => match g {
            Expr::Val(m) => Expr::Val(n + m),
            other => {
                if n == 0 {
                    other
                } else {
                    match other {
                        Expr::Add(left, right) => match (*left, *right) {
                            (Expr::Val(m), tail) => add(Expr::Val(n + m), tail),
                            (head, Expr::Val(m)) => add(Expr::Val(n + m), head),
                            (lhs, rhs) => Expr::Add(
                                Box::new(Expr::Val(n)),
                                Box::new(Expr::Add(Box::new(lhs), Box::new(rhs))),
                            ),
                        },
                        other => Expr::Add(Box::new(Expr::Val(n)), Box::new(other)),
                    }
                }
            }
        },
        Expr::Add(left, right) => add(*left, add(*right, g)),
        other => match g {
            Expr::Val(0) => other,
            Expr::Val(n) => add(Expr::Val(n), other),
            Expr::Add(left, right) => match (*left, *right) {
                (Expr::Val(n), tail) => add(Expr::Val(n), add(other, tail)),
                (head, tail) => Expr::Add(
                    Box::new(other),
                    Box::new(Expr::Add(Box::new(head), Box::new(tail))),
                ),
            },
            rhs => Expr::Add(Box::new(other), Box::new(rhs)),
        },
    }
}

fn mul(f: Expr, g: Expr) -> Expr {
    match f {
        Expr::Val(n) => match g {
            Expr::Val(m) => Expr::Val(n * m),
            other => {
                if n == 0 {
                    Expr::Val(0)
                } else if n == 1 {
                    other
                } else {
                    match other {
                        Expr::Mul(left, right) => match (*left, *right) {
                            (Expr::Val(m), tail) => mul(Expr::Val(n * m), tail),
                            (head, Expr::Val(m)) => mul(Expr::Val(n * m), head),
                            (lhs, rhs) => Expr::Mul(
                                Box::new(Expr::Val(n)),
                                Box::new(Expr::Mul(Box::new(lhs), Box::new(rhs))),
                            ),
                        },
                        other => Expr::Mul(Box::new(Expr::Val(n)), Box::new(other)),
                    }
                }
            }
        },
        Expr::Mul(left, right) => mul(*left, mul(*right, g)),
        other => match g {
            Expr::Val(0) => Expr::Val(0),
            Expr::Val(1) => other,
            Expr::Val(n) => mul(Expr::Val(n), other),
            Expr::Mul(left, right) => match (*left, *right) {
                (Expr::Val(n), tail) => mul(Expr::Val(n), mul(other, tail)),
                (head, tail) => Expr::Mul(
                    Box::new(other),
                    Box::new(Expr::Mul(Box::new(head), Box::new(tail))),
                ),
            },
            rhs => Expr::Mul(Box::new(other), Box::new(rhs)),
        },
    }
}

fn pow(f: Expr, g: Expr) -> Expr {
    match f {
        Expr::Val(m) => match g {
            Expr::Val(n) => Expr::Val(pown(m, n)),
            other => {
                if m == 0 {
                    Expr::Val(0)
                } else {
                    match other {
                        Expr::Val(0) => Expr::Val(1),
                        Expr::Val(1) => Expr::Val(m),
                        other => Expr::Pow(Box::new(Expr::Val(m)), Box::new(other)),
                    }
                }
            }
        },
        other => match g {
            Expr::Val(0) => Expr::Val(1),
            Expr::Val(1) => other,
            rhs => match other {
                Expr::Val(0) => Expr::Val(0),
                lhs => Expr::Pow(Box::new(lhs), Box::new(rhs)),
            },
        },
    }
}

fn ln(f: Expr) -> Expr {
    match f {
        Expr::Val(1) => Expr::Val(0),
        other => Expr::Ln(Box::new(other)),
    }
}

fn deriv(x: i64, expr: Expr) -> Expr {
    match expr {
        Expr::Val(_) => Expr::Val(0),
        Expr::Var(y) => {
            if x == y {
                Expr::Val(1)
            } else {
                Expr::Val(0)
            }
        }
        Expr::Add(f, g) => add(deriv(x, *f), deriv(x, *g)),
        Expr::Mul(f, g) => {
            let f_expr = *f;
            let g_expr = *g;
            add(
                mul(f_expr.clone(), deriv(x, g_expr.clone())),
                mul(g_expr, deriv(x, f_expr)),
            )
        }
        Expr::Pow(f, g) => {
            let f_expr = *f;
            let g_expr = *g;
            mul(
                pow(f_expr.clone(), g_expr.clone()),
                add(
                    mul(
                        mul(g_expr.clone(), deriv(x, f_expr.clone())),
                        pow(f_expr.clone(), Expr::Val(-1)),
                    ),
                    mul(ln(f_expr), deriv(x, g_expr)),
                ),
            )
        }
        Expr::Ln(f) => {
            let f_expr = *f;
            mul(deriv(x, f_expr.clone()), pow(f_expr, Expr::Val(-1)))
        }
    }
}

fn count(expr: &Expr) -> usize {
    match expr {
        Expr::Val(_) | Expr::Var(_) => 1,
        Expr::Add(f, g) | Expr::Mul(f, g) | Expr::Pow(f, g) => count(f) + count(g),
        Expr::Ln(f) => count(f),
    }
}

fn main() {
    let mut expr = pow(Expr::Var(1), Expr::Var(1));
    for step in 0..STEPS {
        expr = deriv(1, expr);
        println!("{} count: {}", step + 1, count(&expr));
    }
}
